contracts/bonding/BondingManager.sol

```diff
@@ -332,68 +335,135 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
         // Update total bonded tokens
-        totalBonded = totalBonded.sub(del.bondedAmount);
-
-        if (transcoderStatus(msg.sender) == TranscoderStatus.Registered) {
-            // If caller is a registered transcoder, resign
-            // In the future, with partial unbonding there would be a check for 0 bonded stake as well
-            resignTranscoder(msg.sender);
+        totalBonded = totalBonded.sub(_amount);
+
+        if (transcoderStatus(del.delegateAddress) == TranscoderStatus.Registered && (del.delegateAddress != msg.sender || del.bondedAmount > 0)) {
+            // A transcoder's delegated stake within the registered pool needs to be decreased if:
+            // - The caller's delegate is a registered transcoder
+            // - Caller is not delegated to self OR caller is delegated to self and has a non-zero bonded amount
+            // If the caller is delegated to self and has a zero bonded amount, it will be removed from the 
+            // transcoder pool so its delegated stake within the pool does not need to be decreased
+            transcoderPool.updateKey(del.delegateAddress, transcoderPool.getKey(del.delegateAddress).sub(_amount), address(0), address(0));
         }
 
-        if (del.delegateAddress != msg.sender && transcoderStatus(del.delegateAddress) == TranscoderStatus.Registered) {
-            // If delegate is not self and is a registered transcoder, decrease its delegated stake
-            // We do not need to decrease delegated stake if delegate is self because we would have already removed self
-            // from the transcoder pool
-            transcoderPool.updateKey(del.delegateAddress, transcoderPool.getKey(del.delegateAddress).sub(del.bondedAmount), address(0), address(0));
+        // Check if delegator is still bonded
+        if (del.bondedAmount == 0) {
+            // Delegator no longer delegated to anyone if it does not have a bonded amount
+            del.delegateAddress = address(0);
+            // Delegator does not have a start round if it is no longer delegated to anyone
+            del.startRound = 0;
+
+            if (transcoderStatus(msg.sender) == TranscoderStatus.Registered) {
+                // If caller is a registered transcoder and is no longer bonded, resign
+                resignTranscoder(msg.sender);
+            }
+        } 
+
+        Unbond(del.delegateAddress, msg.sender, unbondingLockId, _amount, withdrawRound);
+    }
+
+    /**
+     * @dev Rebond tokens for an unbonding lock to a delegator's current delegate
+     * @param _unbondingLockId ID of unbonding lock to rebond with
+     */
+    function rebond(
+        uint256 _unbondingLockId
+    ) 
+        external
+        whenSystemNotPaused
+        currentRoundInitialized 
+        autoClaimEarnings
+    {
+        // Caller must not be an unbonded delegator
+        require(delegatorStatus(msg.sender) != DelegatorStatus.Unbonded);
```

Comment:
What do you think about the addition of a function `rebondFromUnbondingLockWhileUnbonded(uint256 _unbondingLockID)`, which mostly replicates the logic of the bond function, except only does it if the user has an active unbondingLock but is in the Unbonded state?

This would enable rebonding in the case where you're currently Unbonded in the waiting period, but it would keep the logic of the bond() and rebond() methods straightforward without having to handle edge cases?